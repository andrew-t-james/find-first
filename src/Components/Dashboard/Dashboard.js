import React, { Component } from 'react';
import { Route, Switch } from 'react-router-dom';
import Card from '../Card/Card';
import Header from '../Header/Header';
import { connect } from 'react-redux';
import CardContainer from '../CardContainer/CardContainer';
import Footer from '../Footer/Footer';
import Loader from '../Loader/Loader';
import PropTypes from 'prop-types';
import { githubJobsThunk } from '../../thunks/jobListings';

export class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      query: '',
      jobs: this.props.jobs
    };
  }

  componentDidMount = async () => {
    const { jobs } = this.props;

    if (!jobs.length) {
      await this.props.githubJobs();
      this.setState({ jobs: this.props.jobs });
    }
  }

  handleUpdate = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });

    this.filterJobs(value);
  }

  filterJobs = value => {
    const { jobs } = this.state;
    const filteredJobs = this.props.jobs.filter(job => job.title.toLowerCase().includes(value.toLowerCase()));
    this.setState({ jobs: filteredJobs });
  }

  render() {
    const { isLoading } = this.props;
    const { query, jobs } = this.state;

    return (
      <div className="grid-container">
        <Header />
        <main className="content">
          <section className="cards-container">
            <div className="search">
              <input
                type="text"
                value={query}
                name="query"
                onChange={this.handleUpdate}
                className="search__input"
                placeholder="Search Jobs"
              />
            </div>
            {isLoading
              ?
              <div className="loader-container">
                <Loader />
              </div>
              :
              <CardContainer jobs={jobs} />
            }
          </section>
        </main>
        <Footer />
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  jobs: state.githubJobs,
  isLoading: state.isLoading
});

export const mapDispatchToProps = dispatch => ({
  githubJobs: () => dispatch(githubJobsThunk())
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);

Dashboard.propTypes = {

};