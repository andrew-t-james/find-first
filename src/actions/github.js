export const githubJobsAction = jobs => ({
  type: 'GITHUB_JOBS',
  jobs
});