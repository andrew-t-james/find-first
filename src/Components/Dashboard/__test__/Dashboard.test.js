import React from 'react';
import { shallow } from 'enzyme';
import { Dashboard, mapDispatchToProps, mapStateToProps } from '../Dashboard';
import { githubJobsAction } from '../../../Actions/github';

jest.mock('../../../firebase/firebase.js');

describe('<Dashboard />', () => {
  let wrapper;
  const mockGithubJobs = jest.fn();
  const mockGetSavedJobs = jest.fn();
  const mockJobs = [];

  beforeEach(() => wrapper = shallow(
    <Dashboard
      jobs={mockJobs}
      getSavedJobs={mockGetSavedJobs}
      githubJobs={mockGithubJobs}
    />
  ));

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call githubApiRequest on mount', async () => {
    await wrapper.instance().componentDidMount();
    expect(mockGithubJobs).toHaveBeenCalled();
  });

  describe('handleUpdate', () => {
    test('should update state onChange', () => {
      const mockState = {
        query: ''
      };
      const mockEvent = {
        preventDefault: jest.fn(),
        target: {
          name: 'query',
          value: 'my hot job'
        }
      };

      wrapper.instance().handleUpdate(mockEvent);
      expect(wrapper.state('query')).toBe('my hot job');
    });
  });

  describe('filterJobs', () => {
    test('should filter jobs caseInsensitive', () => {
      const mockJobs = [
        {
          title: 'Front End',
          id:1
        },
        {
          title: 'VP Engineering',
          id: 2
        }
      ];
      const expected = [
        {
          title: 'Front End',
          id:1
        }
      ];

      wrapper = shallow(
        <Dashboard
          jobs={mockJobs}
          githubJobs={mockGithubJobs}
        />);

      wrapper.instance().filterJobs('front');
      expect(wrapper.state('jobs')).toEqual(expected);
    });

  });


  describe('mapStateToProps', () => {
    test('should map githubJobs to props', () => {
      const mockState = {
        githubJobs: [
          {
            id: 1,
            title: 'Dev'
          }
        ]
      };

      const expectedState = {
        jobs: [
          {
            id: 1,
            title: 'Dev'
          }
        ]
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expectedState);
    });
  });

  describe('mapDispatchToProps', () => {
    test('should call dispatch when gitHubJobs is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = githubJobsAction();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.githubJobs();
      expect(mockDispatch).toHaveBeenCalled();
    });
  });
});
