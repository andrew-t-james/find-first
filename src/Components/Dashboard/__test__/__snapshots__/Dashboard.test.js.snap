// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> should match snapshot 1`] = `
<e
  closeMenu={[Function]}
  nav={
    Array [
      Object {
        "label": <div
          className="logo"
          id="menu-logo"
        >
          <h1
            className="logo__forward"
          >
            F
          </h1>
          <h1
            className="logo__backward"
          >
            F
          </h1>
        </div>,
        "path": "/",
      },
      Object {
        "id": "sign-in",
        "label": "Sign-In",
        "path": "/sign-in",
      },
      Object {
        "id": "dashboard",
        "label": "Dashboard",
        "path": "/dashboard",
      },
      Object {
        "id": "saved-jobs",
        "label": "Saved Jobs",
        "path": "/saved-jobs",
      },
    ]
  }
  reactRouter={true}
>
  <div
    className="grid-container"
  >
    <Connect(Header) />
    <main
      className="content"
    >
      <section
        className="cards-container"
      >
        <div
          className="search"
        >
          <input
            className="search__input"
            name="query"
            onChange={[Function]}
            placeholder="Search Jobs"
            type="text"
            value=""
          />
        </div>
        <CardContainer
          jobs={Array []}
        />
      </section>
    </main>
    <Connect(Footer) />
  </div>
</e>
`;
